wgpu-core/src/memory_init_tracker.rs

```diff
@@ -0,0 +1,196 @@
+use std::ops::Range;
+
+#[derive(Debug, Clone)]
+pub(crate) enum MemoryInitKind {
+    // The memory range is going to be written by an already initialized source, thus doesn't need extra attention other than marking as initialized.
+    ImplicitlyInitialized,
+    // The memory range is going to be read, therefore needs to ensure prior initialization.
+    NeedsInitializedMemory,
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct MemoryInitTrackerAction<ResourceId> {
+    pub(crate) id: ResourceId,
+    pub(crate) range: Range<wgt::BufferAddress>,
+    pub(crate) kind: MemoryInitKind,
+}
+
+/// Tracks initialization status of a linear range from 0..size
+#[derive(Debug)]
+pub(crate) struct MemoryInitTracker {
+    uninitialized_ranges: Vec<Range<wgt::BufferAddress>>,
+}
+
+pub(crate) struct MemoryInitTrackerDrain<'a> {
+    uninitialized_ranges: &'a mut Vec<Range<wgt::BufferAddress>>,
+    drain_range: Range<wgt::BufferAddress>,
+    next_index: usize,
+}
+
+impl<'a> Iterator for MemoryInitTrackerDrain<'a> {
+    type Item = Range<wgt::BufferAddress>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let uninitialized_range = match self.uninitialized_ranges.get_mut(self.next_index) {
+            Some(range) => range,
+            None => return None,
+        };
+        if uninitialized_range.start >= self.drain_range.end {
+            // No more cuts possible (we're going left to right!)
+            None
+        } else if uninitialized_range.end > self.drain_range.end {
+            // cut-out / split
+            if uninitialized_range.start < self.drain_range.start {
+                let old_start = uninitialized_range.start;
+                uninitialized_range.start = self.drain_range.end;
+                self.uninitialized_ranges
+                    .insert(self.next_index, old_start..self.drain_range.start);
+                self.next_index = std::usize::MAX;
+                Some(self.drain_range.clone())
+            }
+            // right cut
+            else {
+                let result = uninitialized_range.start..self.drain_range.end;
+                self.next_index = std::usize::MAX;
+                uninitialized_range.start = self.drain_range.end;
+                Some(result)
+            }
+        } else {
+            // left cut
+            if uninitialized_range.start < self.drain_range.start {
+                let result = self.drain_range.start..uninitialized_range.end;
+                uninitialized_range.end = self.drain_range.start;
+                self.next_index = self.next_index + 1;
+                Some(result)
+            }
+            // fully contained.
+            else {
+                let result = uninitialized_range.clone();
+                self.uninitialized_ranges.remove(self.next_index);
```

Comment:
O(N^2) again here. We need to think a bit more about the data structures, although we can land this variant for a baseline to improve upon. It's also unfortunate to hit this when all the user is doing being `.for_each(drop)`.

Strictly speaking, I think the whole drain should result in only 1 copy of the data. Instead of moving things one by one, we can do the following:
  - once given a range in `drain_uninitialized_ranges`, we need to find the range of indices in the `uninitialized_ranges` array that are affected.
  - we go through the indices and simply clamp each range to the drain range as we iterate, immutably
  - when the iteration is done, we can then do a single mutation of the `uninitialized_ranges`, where we remove everything from the middle and adjust the border ranges. This is where we may need to insert or remove at most 1 thing.